contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -4453,3 +4464,58 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
 	tsql_outmost_estat->pivot_number++;	
 	MemoryContextSwitchTo(oldContext);
 }
+
+static Node* optimize_explicit_cast(ParseState *pstate, Node *node)
+{
+	if (sql_dialect != SQL_DIALECT_TSQL)
+		return node;
+	if (node == NULL)
+		return NULL;
+	if (IsA(node, FuncExpr))
+	{
+		FuncExpr   *f = (FuncExpr *) node;
+		if (f->funcformat == COERCE_EXPLICIT_CAST){
+			Node* arg = (Node*)linitial(f->args);
+			if (nodeTag(arg) == T_Var && 
+			   ((Var*) arg)->vartype == INT4OID && 
+			   f->funcresulttype == INT8OID)
+				return optimize_explicit_cast(pstate, linitial(f->args));
+		}
+	}
+	else if (IsA(node, BoolExpr))
+	{
+		BoolExpr *r = (BoolExpr *) node;
+		ListCell *l;
+		foreach (l , r->args)
+		{
+			optimize_explicit_cast(pstate, (Node*)lfirst(l));
+		}
+	}
+	else if (IsA(node, OpExpr))
+	{
+		OpExpr *opExpr = (OpExpr*) node;
+		Form_pg_operator form;
+		HeapTuple	tuple;
+		Node* node = optimize_explicit_cast(pstate, linitial(opExpr->args));
+		Node* result = NULL;
+		if (node != linitial(opExpr->args))
+		{
+			char	   *opname;
+
+			tuple = SearchSysCache1(OPEROID, ObjectIdGetDatum(opExpr->opno));
+			if (!HeapTupleIsValid(tuple))
+				return node; // no need to error out in here, stop transform and quite, keep the original node
+			form = (Form_pg_operator) GETSTRUCT(tuple);
+
+			opname = NameStr(form->oprname);
+			if (strcmp(opname, "=") == 0)
```

Comment:
Let's add a comment that currently we only remove unnecessary cast for equal op.